File: server/models/Coupon.js
const mongoose = require('mongoose');

const couponSchema = new mongoose.Schema({
    code: {
        type: String,
        required: true,
        unique: true,
        uppercase: true,
        trim: true
    },
    discountPercent: {
        type: Number,
        required: true,
        min: 1,
        max: 100
    },
    maxDiscount: {
        type: Number,
        default: 0 // 0 means no cap
    },
    usageLimit: {
        type: Number,
        default: 0 // 0 means unlimited
    },
    usedCount: {
        type: Number,
        default: 0
    },
    expiresAt: {
        type: Date,
        default: null
    },
    isActive: {
        type: Boolean,
        default: true
    },
    createdAt: {
        type: Date,
        default: Date.now
    }
});

module.exports = mongoose.model('Coupon', couponSchema);

File: ecom_app/lib/models/coupon.dart
class Coupon {
  final String id;
  final String code;
  final int discountPercent;
  final int maxDiscount;
  final int usageLimit;
  final int usedCount;
  final DateTime? expiresAt;
  final bool isActive;

  Coupon({
    required this.id,
    required this.code,
    required this.discountPercent,
    required this.maxDiscount,
    required this.usageLimit,
    required this.usedCount,
    this.expiresAt,
    required this.isActive,
  });

  factory Coupon.fromJson(Map<String, dynamic> json) {
    return Coupon(
      id: json['_id'] ?? '',
      code: json['code'] ?? '',
      discountPercent: json['discountPercent'] ?? 0,
      maxDiscount: json['maxDiscount'] ?? 0,
      usageLimit: json['usageLimit'] ?? 0,
      usedCount: json['usedCount'] ?? 0,
      expiresAt: json['expiresAt'] != null
          ? DateTime.parse(json['expiresAt'])
          : null,
      isActive: json['isActive'] ?? true,
    );
  }
}
